### PR TITLE
Catch stack overflows when loading from storage

### PR DESCRIPTION
**Pull request by rubenporras**
That is, catch and log StackOverflowErrors when loading resources from
storage, and fall back to recreating the resource from source.
StackOverflowErrors can happen if the AST is deeply nested, since the
algorithm to load from storage uses recursion.

Signed-off-by: rubenporras <Ruben.PorrasCampo@avaloq.com>

### CODE REVIEW

**Comment by rubenporras** [comment]
This is an adapted version of https://github.com/dsldevkit/dsl-devkit/pull/613 for xtext-core.

**Comment by cdietrich** [comment]
@rubenporras do you have an example for the stackoverflow

**Comment by cdietrich** [comment]
so its DirectLinkingResourceStorageLoadable that is recursive?

**Comment by rubenporras** [comment]
Hi @cdietrich,

The recursion is in org.eclipse.emf.ecore.resource.impl.BinaryResourceImpl.EObjectInputStream.loadResource(Resource), called through org.eclipse.xtext.resource.persistence.ResourceStorageLoadable.readContents(...).new EObjectInputStream() {...}.(InputStream, Map<?, ?>).

You can see that org.eclipse.emf.ecore.resource.impl.BinaryResourceImpl.EObjectInputStream.loadEObject() is called recursively, for example through loadEObject -> loadFeatureValue -> loadEObject or loadEObject -> loadFeatureMap -> loadFeatureMapEntry -> loadEObject.

If the user has a file with nested ASTs, and if they are persisted, one will get out of memory when recovering the model.

**Comment by cdietrich** [comment]
hmmm but wont save have the same problem then?

**Comment by rubenporras** [comment]
Hi @cdietrich ,

I have looked at the save implementation, and I think it has also a similar problem. I do not know why we never had problems saving the models, only recovering them. I guess that the loading happens at points where the stack is deeper than at the points where we save them.

Fixing the root problem would require avoiding this type of recursion in the algorithm, or to have a counter that aborts when recursion reaches some level, then one could set Xss accordingly. This PR does not attempt such fix, just a quick-win to make the system as resilient as possible and avoid crashing, given the current algorithm.

If you prefer not to have the safety net and wait for someone to fix it in a better way, I would understand it and just close the PR; we would then keep our "patched" version.

**Comment by cdietrich** [comment]
lets wait what Sebastian thinks

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java [hunk 1]
```diff
@@ -59,6 +82,17 @@ public void load(Map<?, ?> options) throws IOException {
 					eAdapters.clear();
 				}
 				unload();
+			} catch (StackOverflowError e) {
+			      trimStackOverflowErrorStackTrace(e);
+			      LOG.warn("Failed to load " + uri + " from storage", e); //$NON-NLS-1$//$NON-NLS-2$
+
+			      if (contents != null) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java:89]: Nit: please extract a protected method `unload(Throwable t)`?  that's used from both catch blocks.

Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java:89]: Maybe call it `loadFromStorageFailed(Throwable)`

Comment (rubenporras) [reply org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java]: @szarnekow , done

**Review by szarnekow — APPROVED** [review]

**Comment by rubenporras** [comment]
@cdietrich , could we merge? Thanks.

**Pull request merged**